src/Uno.UI.Runtime.Skia.X11/X11PointerInputSource.XInput.cs

```diff
@@ -0,0 +1,513 @@
+			timestamp: (ulong)data.time,
+			data.evtype is XiEventType.XI_TouchBegin or XiEventType.XI_TouchEnd or XiEventType.XI_TouchUpdate ? PointerDevice.For(PointerDeviceType.Touch) : PointerDevice.For(PointerDeviceType.Mouse),
+			(uint)data.sourceid,
+			new Point(data.event_x / scale, data.event_y / scale),
+			new Point(data.event_x / scale, data.event_y / scale),
+			properties.HasPressedButton,
+			properties
+		);
+
+		var modifiers = X11XamlRootHost.XModifierMaskToVirtualKeyModifiers((XModifierMask)(data.mods.Base & 0xffff));
+
+		return new PointerEventArgs(point, modifiers);
+	}
+
+	// This method is comment free. See the very similar (but more involved) implementation of
+	// CreatePointerEventArgsFromDeviceEvent for comments.
+	public unsafe PointerEventArgs CreatePointerEventArgsFromEnterLeaveEvent(XIEnterLeaveEvent data, PointerDeviceType pointerType)
+	{
+		var mask = 0;
+		for (var i = 0; i < data.buttons.MaskLen; i++) // masklen <= 4
+		{
+			mask |= data.buttons.Mask[i] << (8 * i);
+		}
+
+		var properties = new PointerPointProperties
+		{
+			IsLeftButtonPressed = (mask & (1 << LEFT)) != 0,
+			IsMiddleButtonPressed = (mask & (1 << MIDDLE)) != 0,
+			IsRightButtonPressed = (mask & (1 << RIGHT)) != 0,
+			IsXButton1Pressed = (mask & (1 << XButton1)) != 0,
+			IsXButton2Pressed = (mask & (1 << XButton2)) != 0,
+			IsInRange = true,
+			IsTouchPad = GetDevicePropertiesFromId(data.sourceid) is { } info && IsTouchpad(info.Properties),
+			IsHorizontalMouseWheel = false,
+		};
+
+		var point = new PointerPoint(
+			frameId: (uint)data.time, // UNO TODO: How should we set the frame, timestamp may overflow.
+			timestamp: (ulong)data.time,
+			PointerDevice.For(PointerDeviceType.Mouse),
+			(uint)data.sourceid,
+			new Point(data.event_x, data.event_y),
+			new Point(data.event_x, data.event_y),
+			properties.HasPressedButton,
+			properties
+		);
+
+		var modifiers = X11XamlRootHost.XModifierMaskToVirtualKeyModifiers((XModifierMask)(data.mods.Base & 0xffff));
+
+		return new PointerEventArgs(point, modifiers);
+	}
+
+	// Note about removing devices: the server emits a ButtonRelease if a device is removed
+	// while a button is held.
+	public void HandleXI2Event(XEvent ev)
+	{
+		var evtype = (XiEventType)ev.GenericEventCookie.evtype;
+		if (this.Log().IsEnabled(LogLevel.Trace))
+		{
+			this.Log().Trace($"XI2 EVENT: {evtype}");
+		}
+		switch (evtype)
+		{
+			case XiEventType.XI_Enter:
+			case XiEventType.XI_Leave:
+				{
+					_lastHorizontalTouchpadWheelPosition.Clear();
+					_lastVerticalTouchpadWheelPosition.Clear();
+					var args = CreatePointerEventArgsFromEnterLeaveEvent(
+						ev.GenericEventCookie.GetEvent<XIEnterLeaveEvent>(),
+						PointerDeviceType.Mouse);
+					if (evtype is XiEventType.XI_Enter)
+					{
+						X11XamlRootHost.QueueAction(_host, () => RaisePointerEntered(args));
+					}
+					else
+					{
+						X11XamlRootHost.QueueAction(_host, () => RaisePointerExited(args));
```

Comment:
Should it check `XiEventType.detail` to `XiEnterLeaveDetail.XINotifyAncestor` here? The XINotifyAncestor means the application receive the leave from the ancestor window, such as dialog.

Reply:
We currently never use an Uno window as a modal/dialog window. But noted for future reference. Thanks!